From 6f46a492048865e75ff9c587090271d47181c11a Mon Sep 17 00:00:00 2001
From: Hongzhang Yang <Hongzhang.Yang@freescale.com>
Date: Fri, 1 Feb 2013 16:59:05 +0800
Subject: [PATCH 386/505] ENGR00242830 Correct field type of frame buffer to
 be passed to V4L QBUF

Bug:
Field type of frame buffer to be passed to V4L QBUF does not match with
the content.

Root cause:
- Doesn't set V4L2_FIELD_NONE to progressive frame
- field passed to VIDIOC_QBUF is the type of decoded buffer, not the
  display buffer

BTW, move COMMON_INIT to vpu_test.h to affect all files

Signed-off-by: Hongzhang Yang <Hongzhang.Yang@freescale.com>
---
 test/mxc_vpu_test/dec.c      | 9 ++++++---
 test/mxc_vpu_test/main.c     | 1 -
 test/mxc_vpu_test/vpu_test.h | 6 +++++-
 3 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/test/mxc_vpu_test/dec.c b/test/mxc_vpu_test/dec.c
index ab2641f..102fca4 100644
--- a/test/mxc_vpu_test/dec.c
+++ b/test/mxc_vpu_test/dec.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2004-2012 Freescale Semiconductor, Inc.
+ * Copyright 2004-2013 Freescale Semiconductor, Inc.
  *
  * Copyright (c) 2006, Chips & Media.  All rights reserved.
  */
@@ -1133,6 +1133,8 @@ decoder_start(struct decode *dec)
 		}
 
 		if (outinfo.indexFrameDecoded >= 0) {
+			field = V4L2_FIELD_NONE;
+
 			if (dec->cmdl->format == STD_VC1) {
 				if (outinfo.pictureStructure == 2)
 					info_msg("dec_idx %d : FRAME_INTERLACE\n", decIndex);
@@ -1188,6 +1190,7 @@ decoder_start(struct decode *dec)
 				if (outinfo.repeatFirstField)
 					info_msg("frame_idx %d : Repeat First Field\n", decIndex);
 			}
+			dec->decoded_field[outinfo.indexFrameDecoded]= field;
 		}
 
 		if(outinfo.indexFrameDecoded >= 0)
@@ -1274,9 +1277,9 @@ decoder_start(struct decode *dec)
 
 			if (!cpu_is_mx27())
 				if (dec->cmdl->dst_scheme == PATH_V4L2)
-					err = v4l_put_data(dec, actual_display_index, field, dec->cmdl->fps);
+					err = v4l_put_data(dec, actual_display_index, dec->decoded_field[actual_display_index], dec->cmdl->fps);
 				else
-					err = ipu_put_data(disp, actual_display_index, field, dec->cmdl->fps);
+					err = ipu_put_data(disp, actual_display_index, dec->decoded_field[actual_display_index], dec->cmdl->fps);
 			else
 				if (dec->cmdl->dst_scheme == PATH_V4L2)
 					err = v4l_put_data(dec, actual_display_index, V4L2_FIELD_ANY, dec->cmdl->fps);
diff --git a/test/mxc_vpu_test/main.c b/test/mxc_vpu_test/main.c
index 2e57889..775eb50 100644
--- a/test/mxc_vpu_test/main.c
+++ b/test/mxc_vpu_test/main.c
@@ -25,7 +25,6 @@
 #include <getopt.h>
 #include "vpu_test.h"
 
-#define COMMON_INIT
 #define ONE_FRAME_INTERV 100000 // 100 ms
 
 char *usage = "Usage: ./mxc_vpu_test.out -D \"<decode options>\" "\
diff --git a/test/mxc_vpu_test/vpu_test.h b/test/mxc_vpu_test/vpu_test.h
index cf5bdea..3a201c6 100644
--- a/test/mxc_vpu_test/vpu_test.h
+++ b/test/mxc_vpu_test/vpu_test.h
@@ -1,5 +1,5 @@
 /*
- * Copyright 2004-2012 Freescale Semiconductor, Inc.
+ * Copyright 2004-2013 Freescale Semiconductor, Inc.
  *
  * Copyright (c) 2006, Chips & Media.  All rights reserved.
  */
@@ -25,6 +25,8 @@
 #include "vpu_lib.h"
 #include "vpu_io.h"
 
+#define COMMON_INIT
+
 extern int vpu_test_dbg_level;
 
 #define dprintf(level, fmt, arg...)     if (vpu_test_dbg_level >= level) \
@@ -226,6 +228,8 @@ struct decode {
 	DecReportInfo userData;
 
 	struct cmd_line *cmdl;
+
+	int decoded_field[32];
 };
 
 struct encode {
-- 
1.8.0

